File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarSkinTemperatureUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.PolarSkinTemperatureResult
import com.polar.sdk.api.model.SkinTemperatureMeasurementType
import com.polar.sdk.api.model.SkinTemperatureSensorLocation
import com.polar.sdk.api.model.fromPbTemperatureMeasurementSamples
import com.polar.services.datamodels.protobuf.TemperatureMeasurement.TemperatureMeasurementPeriod
import protocol.PftpRequest
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

private const val ARABICA_USER_ROOT_FOLDER = "/U/0/"
private const val SKIN_TEMPERATURE_DIRECTORY = "SKINTEMP/"
private const val SKIN_TEMPERATURE_PROTO = "TEMPCONT.BPB"
private val dateFormatter = DateTimeFormatter.ofPattern("yyyyMMdd", Locale.ENGLISH)
private const val TAG = "PolarSkinTemperatureUtils"

internal object PolarSkinTemperatureUtils {

    /**
     * Read skin temperature data for a given date.
     */
    suspend fun readSkinTemperatureDataFromDayDirectory(client: BlePsFtpClient, date: LocalDate): PolarSkinTemperatureResult? {
        BleLogger.d(TAG, "readSkinTemperatureDataFromDayDirectory: $date")
        val skinTempFilePath = "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$SKIN_TEMPERATURE_DIRECTORY$SKIN_TEMPERATURE_PROTO"
        return try {
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(skinTempFilePath)
                    .build()
                    .toByteArray()
            )
            val proto = TemperatureMeasurementPeriod.parseFrom(response.toByteArray())
            PolarSkinTemperatureResult(
                proto.sourceDeviceId,
                SkinTemperatureSensorLocation.from(proto.sensorLocation.ordinal),
                SkinTemperatureMeasurementType.from(proto.measurementType.ordinal),
                fromPbTemperatureMeasurementSamples(proto.temperatureMeasurementSamplesList)
            )
        } catch (error: Throwable) {
            BleLogger.w(TAG, "Failed to fetch skin temperature data for date: $date, error: $error")
            null
        }
    }
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarSkinTemperatureUtils.swift
//  Copyright © 2025 Polar. All rights reserved.

import Foundation

private let ARABICA_USER_ROOT_FOLDER = "/U/0/"
private let SKIN_TEMPERATURE_DIRECTORY = "SKINTEMP/"
private let SKIN_TEMPERATURE_PROTO = "TEMPCONT.BPB"
private let dateFormat: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyyMMdd"
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = TimeZone(abbreviation: "UTC")
    return formatter
}()
private let TAG = "PolarSkinTemperatureUtils"

internal class PolarSkinTemperatureUtils {
    static func readSkinTemperatureData(client: BlePsFtpClient, date: Date) async -> PolarSkinTemperatureData.PolarSkinTemperatureResult? {
        let dateString = dateFormat.string(from: date).components(separatedBy: "±").first ?? ""
        BleLogger.trace(TAG, "readSkinTemperatureData: \(dateString)")
        let filePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateString)/\(SKIN_TEMPERATURE_DIRECTORY)\(SKIN_TEMPERATURE_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = filePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let skinTemp = try Data_TemperatureMeasurementPeriod(serializedBytes: Data(response))
            return PolarSkinTemperatureData.PolarSkinTemperatureResult(
                date: date,
                sensorLocation: PolarSkinTemperatureData.SkinTemperatureSensorLocation.getByValue(value: skinTemp.sensorLocation),
                measurementType: PolarSkinTemperatureData.SkinTemperatureMeasurementType.getByValue(value: skinTemp.measurementType),
                skinTemperatureList: PolarSkinTemperatureData.fromPbTemperatureMeasurementSamples(pbTemperatureMeasurementData: skinTemp.temperatureMeasurementSamples)
            )
        } catch {
            BleLogger.error("readSkinTemperatureData() failed for path: \(filePath), error: \(error)")
            return nil
        }
    }
}
